### ISSUE TITLE
git dependencies should allow semver requirement

### ISSUE THREAD

**Issue by johnslemmer** [python-poetry/poetry#925]
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://poetry.eustace.io/docs/) and believe that my question is not covered.

## Feature Request

First thanks for creating this much needed tool for the python community.

I don't see this in the documentation or an existing issue for it, but it would be awesome if I could have a git dependency something like this:
```
requests = { git = "https://github.com/kennethreitz/requests.git", semver = "^2.20.0" }
```

That would allow for automatic updating based on the git tags that meet the semver requirement and match this regex in their tag name: `^v?\d+\.\d+\.\d+$` (or something similar).

As an example, NPM allows for this https://docs.npmjs.com/files/package.json#git-urls-as-dependencies

**Comment by darrenleeweber** [comment]
Tested using a tag like `tag = "^0.2"` on a repository with sem-ver tags matching `^v?\d+\.\d+\.\d+$` (this repo actually has no `v` prefix on the versions and it has exiting 0.2.x sem-ver tags).  The current poetry update does not search the tags for a matching sem-ver tag to checkout, it assumes a literal tag to checkout and fails, e.g.

```
$ poetry update
Updating dependencies
Resolving dependencies... (4.1s)
                                                                                                                         
[CalledProcessError]                                                                                      
Command '['git', '--git-dir', '/tmp/pypoetry-git-jupiter-log740h8tq9/.git', '--work-tree', '/tmp/pypoetry-git-jupiter-log740h8tq9', 'checkout'  
, '^0.2']' returned non-zero exit status 1.
```

**Comment by darrenleeweber** [comment]
PR #1176 is up for review

**Comment by darrenleeweber** [comment]
@sdispater notes a veto-vote on #1176 (in it's current form) and proposes that a plugin solution (#1237) _might_ fly.  I'm considering #1176 as a WIP that could work but might impose specialized use-cases on poetry, although the implementation should only kick-in if/when the git-tag specifier is a constraint.  Note that it essentially modifies one function in the git-clone-checkout workflow to select a git-tag.  As a plugin, it would not constitute any new CLI commands and it's not obvious how many events could/should trigger it.  If it is triggered, it's like a monkey-patch on the git-method could be invoked (but not a great idea), so it might amount to using a sub-class of the existing git-class or some form of factory for git behavior or dependency injection.  At this time, I'm inclined to try to consolidate some work on #1176 into a new branch (new PR) with a git-semver subclass or alternate class, to totally isolate the option from the mainstream workflow(s).  Work on any plugin for this obviously depends on plugin infrastructure (slated post 1.0) and clarification for the design for this feature as a plugin.

**Comment by darrenleeweber** [comment]
This depends on or is blocked by plugin feature(s).

**Comment by finswimmer** [comment]
Hello,

I'm not aware this is something that could be translated to PEP 508. But this is necessary to build properly wheels.

fin swimmer

**Comment by jbkoh** [comment]
Hi all, I just want to check in if it's something that poetry could potentially pursue or not. I'm heavily using github tag as a semver of my private git repo other than having a separate python package host like a private PyPI server. If poetry can support interpreting git tags as semver, it'd be a huge help. 

Actually semver vs pep404 doesn't matter to me but rather I would like to use git tag to represent its version. Let me know if a solution already exists.

Thanks a lot.

**Comment by emrojo** [comment]
Hi, does anybody know if there is a way of setting restrictions on version numbers (like ~=1.0.0) when using a private git with no package repository, only tags stored in there? I found this is supported with other package managers like npm or php composer
